### ISSUE TITLE
evaluator: fix compiler paths

### ISSUE THREAD

**Issue by jdegoes** [slamdata/slamengine#659]
The query must be parsed, and using the paths inside, the appropriate backend must be located, and then the query must be re-written to use the relative path from that backend.

**Comment by mossprescott** [comment]
@sellout: if you're planning to work on this, let's talk first because it will interact with what I'm doing in #674.

**Comment by sellout** [comment]
@mossprescott – I guess I’m planning on it now. #674 is done, but I’d like to go over this code, as I’ve never looked at the mounting stuff at all. Monday?

**Comment by jdegoes** [comment]
It's a horrible mess. Here's where the notion of nested file system and unification of backend and file system can help: a nested file system is given an `exec` order to "execute" a query relative to some base path. It parses the query and determines if it can provide the data. If so, it executes the query. If it cannot provide the data, it looks for the mount(s) that can, and relativizes the query and the base path to that level of the hierarchical file system. It then sends the query' to the mount that can handle this, which then repeats the process.

Basically each file system thinks it is the root, and can only ever handle queries relative to the root, but file systems that support mounting can relativize a query together with its base path, and pass along both to the child file systems.

This is the mechanism we'd use for distributed queries, with a file system that supports distribution: it'd exec sub-queries to each backend and finish up computation in one or the other or client-side.

**Comment by mossprescott** [comment]
@jdegoes: a couple of clarifying quesitons:

How will this interact with configuration? I guess we read the flat configuration, figure out which mount paths are nested within one another, and then construct the tree of filesystems. I don't think we want to mess with the configuration abstraction at this point, do we?

Is there a motivating use case right now? Getting rid of the need to specify a path within the correct mount seems like a plus. For instance, you could start a REPL and immediately run `select * from "/local/test/zips"` without first doing `cd local/test`. Presumably the client will benefit from that as well?

**Comment by jdegoes** [comment]
>  I don't think we want to mess with the configuration abstraction at this point, do we?

Probably not given the impending deadline.

> Getting rid of the need to specify a path within the correct mount seems like a plus. For instance, you could start a REPL and immediately run select * from "/local/test/zips" without first doing cd local/test. Presumably the client will benefit from that as well?

Yes, that's correct. The UI will not expose any notion of "query from within this mount", and exposing such a notion would be painful and confusing. So we won't do it.

I consider this a bug because the behavior was always broken (IMO). When you query, the source files specified in `FROM`, `JOIN`, and `CROSS` clauses should have paths which are relative to the base path at which the query is executed, which may be root or may be something else entirely.

Thus, queries from inside any notebook can reference any data source anywhere; in general these queries will use absolute paths. For intermediate data sets, which are stored inside the notebook itself, they may use relative paths (e.g. `SELECT * FROM out1`), because the notebook will evaluate queries relative to the directory containing the notebook. This is done to provide a clean way of deleting all intermediate data sets merely by deleting the notebook itself.

Does this make sense?

**Comment by mossprescott** [comment]
I have a medium-sized refactoring underway that's looking pretty good so far. Hoping to have it solid in time to merge this evening.

**Comment by mossprescott** [comment]
Arg. After a long day, I have everything working (REPL, integration tests) except error cases in the API, but those cases are turning out to require non-trivial refactoring. Basically, the API used to manage the mapping of paths to backends, and handled the errors that occurred directly. Now that mapping is done inside `Backend`, which is much cleaner, but it means that errors need to be represented in `Backend`'s method signatures in a way that they currently aren't.

Given the complexity of the changes, I think there's significant risk of breaking the API in some subtle way. Add to that the fact that this ticket isn't causing any issues for the front-end as far as I know, and I'm now thinking it should be pushed to post-2.0.

**Comment by jdegoes** [comment]
If I understand this correctly (I know the code changed somewhat when we started mounting clusters and not databases), the only case in which this would *not* affect the front-end is the case in which a MongoDB cluster is mounted to root. In this case,, the paths in the queries would be relative to the mount, because the mount would be root. In all other cases, clusters users will encounter this issue.

**Comment by sellout** [comment]
Is this what slamdata/slamdata#196 is about?

**Comment by jdegoes** [comment]
@sellout Yep, which is why this needs to be in ASAP. :-/

**Comment by jdegoes** [comment]
@mossprescott This is still a high priority as all clusters mounted to non-root will not be queryable. Can you make sure to push your branch to Github so @sellout can take a look tomorrow if necessary?

**Comment by sellout** [comment]
@mossprescott did push his branch. I'm working on this, but I seem to have no wifi at the hotel now, so I'm about to relocate.

**Comment by mossprescott** [comment]
@sellout: if you do dig into this, you should know:

The `Backend` signatures need to be reworked. Returning path errors from NestedBackend in the resulting Task/Process is causing too much pain for the API services. I think it would be better to have all those methods return `PathError \/ Task[...]`, so that the service could immediately handle path problems and only `delay` the actual execution. Note the many API tests currently failing on that branch.

This branch does not yet solve the problem of mounts not at the root. I think we just need some code to construct a NestedBackend for each directory in the path, and put the mounts in the right places. It would great to be able to unit test that; maybe we need a null/stub backend for that.

**Comment by jdegoes** [comment]
> I think it would be better to have all those methods return PathError \/ Task[...]

`ErrorT[Task, PathError, ?]`

**Comment by mossprescott** [comment]
Wouldn't that put the error inside the `Task`? That's the opposite of what we need. And I don't think `TaskT` is a thing, is it?

**Comment by jdegoes** [comment]
>  That's the opposite of what we need

Why? Task isn't necessarily asynchronous. You can use Task.now for pure values.

TaskT cannot exist (although theoretically a TaskT-like thing could be stacked on any monad at least as or more powerful than Task to produce another monad).

**Comment by mossprescott** [comment]
I need to understand these trade-offs better.

The way it's working now is these errors come back via `Task.fail`, which means you have to do Task stuff to handle them (`attempt`, e.g.). It also means that the errors are untyped (because `Task.fail` takes a `Throwable`. Also, a couple of the methods actually return `Process[Task, ...]` rather than Task, and in that case failure works a different way. 

My feeling was that you either pass validation and attempt to do some operation on the backend (which will have to be in `Task` or `Process`), or you fail during validation (of request parameters, or of paths, or by attempting some operation that's not supported by a particular backend, etc.) So putting Task/Process inside the `\/` seemed to be a way to capture that.

On the other hand, maybe stacking `ErrorT` on Task/Process will simplify things enough to be a good solution as well.

**Comment by sellout** [comment]
I do think that “logically”, having things like PathError on the outside makes sense, since it indicates that we fail before creating the Task. But I figured `EitherT` would be a bit simpler. Having done most of it with `EitherT` I’m not sure it actually turned out that way.

**Comment by jdegoes** [comment]
> I do think that “logically”, having things like PathError on the outside makes sense, since it indicates that we fail before creating the Task. 

But the thing you end up with is not a monad, it's not anything, and it doesn't compose in all the standard, well-known ways. Also, the first time you have to check a path after some effect / async action, you'll kick yourself for having to report a `PathError` as a `Throwable`. 

`Task`'s error channel can be abused (as we have demonstrated :smiling_imp: ). It should really only be used for fatal and Java errors. Our ours should we well-typed ADTs layered onto `Task`.

Actually, if you look it up, `Task` is just `Future[E \/ A]` for some arbitrary error type, which is equivalent to `EitherT[Future, E, A]`.
